src/controllers/orderController.js

```diff
@@ -5,58 +5,103 @@ const { BadRequestError } = require('../errors');
 const PurchasedAlbum = require('../models/PurchasedAlbum');
 const CustomError = require('../errors');
 const { checkPermissions } = require('../utils');
+const { sendOrderCompletedEmail } = require('../mailing/sender');
+const mongoose = require('mongoose');
 
 const createOrder = async (req, res) => {
-  const { orderItems, subtotal, tax, total } = req.body;
+  const session = await mongoose.startSession();
+
+  try {
+    session.startTransaction();
+
+    const { orderItems, subtotal, tax, total } = req.body;
+
+    if (!orderItems?.length) {
+      throw new BadRequestError(
+        'Unable to create an order because the order items are missing.'
+      );
+    }
+    // Fetch user information from the request object
+    const user = req.user;
 
-  // Perform necessary checks and validations
-  if (!orderItems?.length) {
-    throw new BadRequestError(
-      'Unable to create order because the order items are missing.'
-    );
-  }
 
-  // Create an Order document in the database
-  const order = await Order.create({
-    user: req.user.userId,
-    orderItems,
-    subtotal,
-    tax,
-    total,
-  });
 
-  // Create a payment intent with Stripe
-  const paymentIntent = await stripe.paymentIntents.create({
-    amount: total * 100, // Stripe requires amount in cents
-    currency: 'usd',
-    description: `Order #${order._id}`,
-    metadata: {
-      userId: req.user.userId,
-      orderId: order._id, // key-values
-      totalQuantity: orderItems.reduce(
-        (total, item) => total + item.quantity,
-        0
-      ), // the total number of items in the order
-      totalAlbums: orderItems.length, // the total number of albums purchased in order
-    },
-  });
 
-  // Save the paymentIntent to the Order model
-  order.paymentIntentId = paymentIntent.id;
-  await order.save();
 
-  // Create PurchasedAlbum entries for each album in orderItems
-  for (const orderItem of orderItems) {
-    await PurchasedAlbum.create({
-      album: orderItem.album, // album ID from orderItem
-      user: req.user.userId, // user ID from the request
+// Populate the orderItems with full album data
+const populatedOrderItems = await Promise.all(
+  orderItems.map(async (orderItem) => {
+    return await orderItem.populate('album').execPopulate();
+  })
+);
```

Comment:
Currently this won't work because `populate` is a mongoose function that works on mongoose model instances. But the `orderItems` variable isn't one we got from the database, it's a plain Javascript object that we got from the HTTP request.